kernel/libs/aster-bigtcp/src/socket/bound.rs

```diff
@@ -79,6 +83,22 @@ impl DerefMut for RawTcpSocketExt {
     }
 }
 
+impl RawTcpSocketExt {
+    fn on_new_state(&mut self) -> SocketEvents {
```

Comment:
> This method may be better renamed as `calc_socket_events`.

Calculating the events _to notify_ is NOT  calculating the events. The semantics here are different from `calc_io_events`, so we'd better not choose a similar name.

Also, this method will not be called if the TCP state is not changing (which should be the normal case when transferring a large amount of data). So this is a slow path. (Although I admit that the computation here is not very heavy.)

> Why `SocketEvents::CAN_SEND` and `SocketEvents::CAN_RECV` is not calculated here, but in a seperate place?

They should be in the fast path.

Reply:
> Also, this method will not be called if the TCP state is not changing (which should be the normal case when transferring a large amount of data). So this is a slow path. (Although I admit that the computation here is not very heavy.)

I agree that it deserves a different name. 

The reason I felt strange at first is that checking the SEND and RECEIVE states is inexpensive, yet it needs to be done every time a packet is sent or received, not just when there's a change in state. However, having a fast path is certainly welcome. Overall, I think the current approach is acceptable.

Reply:
To be honest, checking `can_send()`/`can_recv()` is also kind of wired, because:
 - We should only notify `CAN_SEND` when the ACK packet frees up _new_ space, and notify `CAN_RECV` when the data packet brings _new_ data.
 - But this is different from checking for free space and data to receive (which can come from an existing state rather than the state change caused by the _new_ packet).

The best place to do this is probably inside smoltcp's socket implementation, since changing it requires much more effort, so I'll just fall back on a simplest solution that always notifies `CAN_SEND` and `CAN_RECV`.